src/iterators/intersection_iterator.c

```diff
@@ -0,0 +1,294 @@
+}
+
+static inline IteratorStatus II_ReadFromFirstChild(IntersectionIterator *it, t_docId *out) {
+  // Read from the first child iterator, which is guaranteed to be non-NULL
+  RS_ASSERT(it->num_its > 0 && it->its[0] != NULL);
+  QueryIterator *child = it->its[0];
+  IteratorStatus rc = child->Read(child);
+  if (rc == ITERATOR_OK) {
+    *out = child->lastDocId; // If we read successfully, we return the docId
+  } else if (rc == ITERATOR_EOF) {
+    it->base.atEOF = true; // If the first child is at EOF, the intersection is also at EOF
+  }
+  return rc;
+}
+
+/**
+ * Check if all iterators agree on the current docId `curTarget` holds.
+ * If they do, aggregate their results into `current` and return ITERATOR_OK.
+ * If any of the iterators is at EOF, set `atEOF` to true and return ITERATOR_EOF.
+ * If any of the iterators is not at the requested docId, advance it to the requested `docId` and
+ * return ITERATOR_NOTFOUND. The caller may retry calling this function in that case.
+ * In case of an error, return the error code.
+ */
+static IteratorStatus II_AgreeOnDocId(IntersectionIterator *it, t_docId *curTarget) {
+  const t_docId docId = *curTarget;
+
+  for (uint32_t i = 0; i < it->num_its; i++) {
+    RS_ASSERT(it->its[i]->lastDocId <= docId);
+    if (it->its[i]->lastDocId < docId) {
+      // Advance the iterator to the requested docId
+      IteratorStatus rc = it->its[i]->SkipTo(it->its[i], docId);
+      if (rc != ITERATOR_OK) { // Not OK, we break out of the loop
+        if (rc == ITERATOR_EOF) {
+          // Some child iterator reached EOF, so the intersection is also at EOF
+          it->base.atEOF = true;
+        } else if (rc == ITERATOR_NOTFOUND) {
+          // The child iterator did not find the requested docId, so we need to advance the lastDocId
+          // to the next possible value (the result that the child iterator yielded)
+          *curTarget = it->its[i]->lastDocId;
+        }
+        return rc;
+      }
+    }
+  }
+  // All iterators agree on the docId, so we can set the current result
+  AggregateResult_Reset(it->base.current);
+  for (uint32_t i = 0; i < it->num_its; i++) {
+    RS_ASSERT(docId == it->its[i]->current->docId);
+    AggregateResult_AddChild(it->base.current, it->its[i]->current);
+  }
+  return ITERATOR_OK;
+}
+
+static inline IteratorStatus II_Read_Internal_CheckRelevancy(IntersectionIterator *it, t_docId docId) {
+  IteratorStatus rc;
+  do { // retry until we agree on the docId
+    rc = II_AgreeOnDocId(it, &docId);
+    if (rc != ITERATOR_OK) {
+      continue;
+    }
+    if (!II_currentIsRelevant(it)) {
+      rc = II_ReadFromFirstChild(it, &docId);
+      continue;
+    }
+    // Hit!
+    it->base.lastDocId = it->base.current->docId;
+    return ITERATOR_OK;
+
+  } while (rc == ITERATOR_OK || rc == ITERATOR_NOTFOUND);
+  return rc;
+}
+
+static inline IteratorStatus II_Read_Internal(IntersectionIterator *it, t_docId docId) {
```

Comment:
For better understanding of the code, would `II_Find_Consensus` be ab better name? (in counter to the ReadFirstChild)